src/tree/gpu_hist/histogram.cu

```diff
@@ -105,20 +105,27 @@ __global__ void SharedMemHistKernel(EllpackDeviceAccessor matrix,
                                     common::Span<const RowPartitioner::RowIndexT> d_ridx,
                                     GradientSumT* __restrict__ d_node_hist,
                                     const GradientPair* __restrict__ d_gpair,
-                                    size_t n_elements,
+                                    const int* feature_groups, const int* bin_groups,
                                     GradientSumT const rounding,
                                     bool use_shared_memory_histograms) {
   using T = typename GradientSumT::ValueT;
   extern __shared__ char smem[];
+  int group = blockIdx.y;
```

Comment:
I would like kernel code to be much more readable. Some suggestions:
- add comments defining what is a bin_group/feature_group
- add comments talking about your parallel decomposition of work e.g. how each thread block is given a submatrix from the original matrix, and that submatrix may have the width of the original matrix if the data is sparse or has few features

You could create a struct called BlockWorkDecomposition that contains start/end rows and features and move some of this code to its constructor.

Imagine you are trying to convey what is happening here to a new xgboost developer.

Reply:
Now that there is a separate data type for feature groups, I think the code is more readable.